File: practicas/Semana-05/PiPorSeriesThreads.py
'''
   Este programa PiPorSerieThreads lleva a cabo calculos de aproximación de PI usando threads
   Author	Programacion Concurrente
   Version	2020/Abr/11
        __
       \
  pi =  >  (-1)^i x 4/(2xi + 1)
       /__
       i=0

'''

import sys
import threading

'''
   Esta clase valores es necesaria para obtener los valores de la suma de los terminos de PI que realiza cada hilo
   En Python los parametros de los metodos son por valor y no pueden ser cambiados
'''
class valores:
	pass

# Cada hilo ejecuta este metodo que realiza la suma parcial de sus terminos y acumula el resultado en la instancia "total"
def calcula( inicio, terminos, total ):
	miSuma=0.0
	alterna = 4.0 if ( 0 == (inicio % 2) ) else -4.0
	for i in range( int(terminos) ):
		pos = i + inicio
		miSuma += alterna/(2 * pos + 1)
		alterna *= -1.0
	total.suma += miSuma


def main():
	threads = []
	valor = valores
	valor.suma = 0
	terminos = 1000000
	if ( len( sys.argv ) > 1 ):
		terminos = int( sys.argv[ 1 ] )
	else:
		print( "Uso: " + "\033[1m" + sys.argv[ 0 ] + " n (para n un entero)" + "\033[0m" )
	fin = terminos/10
	x = float( 0 )
	for i in range( 10 ):
		inicio = i * terminos/10
		t = threading.Thread( target=calcula, args=( inicio, fin, valor ) )
		threads.append( t )
		t.start()

	for t in threads:
		t.join()

	print( "Valor aproximado de PI: " + '\033[91m' + str( valor.suma ) + '\033[0m' + " para " + str( terminos ) + " elementos")


if __name__ == "__main__":
	main()

File: practicas/Semana-05/PiPorSeriesThreads.java
/**
 * La clase PiPorSerieThreads lleva a cabo calculos de aproximación de PI usando threads
 *
 * @author Programacion Concurrente
 * @version 2020/Abr/11
 *       __
 *      \
 * pi =  >  (-1)^i x 4/(2xi + 1)
 *      /__
 *      i=0
 *
 */

/**
 * Esta clase Suma es necesaria para obtener los valores de la suma de los terminos de PI
 * que realiza cada hilo
 * En Java, existen varias maneras de crear hilos de ejecucion, una de ellas es implantando
 * la clase "Runnable" donde el metodo "run" ya esta predefinido, void, sin parametros y por
 * lo tanto no puede retornar resultados,
 *
 * Cada hilo crea una instancia de esta clase para almacenar la suma de los terminos que le
 * toco calcular
 */
class Suma {
    private double suma;

    public Suma() {
       this.suma = 0;
    }

    public void setSuma( double suma ) {
       this.suma = suma;
    }

    public double getSuma() {
       return this.suma;
    }

}

/**
 * Esta es la clase que vamos a correr cada vez que se crea un hilo
 * Se utiliza el metodo "run" para realizar la tarea de cada hilo
 *
 */
class SeriePorPartes implements Runnable {
    private long fin;
    private long inicio;
    private Suma miParte;

/**
 * En el constructor se definen los limites para los que este hilo realiza
 * el calculo de terminos
 * @param	long	inicio, es el punto donde inicia la suma que calcula este hilo
 * @param	long	fin, es la cantidad de terminos que le toca calcular a este hilo
 * @param	Suma	miParte, una referencia de Suma para guardar el resultado de este hilo
 */
    public SeriePorPartes( long inicio, long fin, Suma miParte ) {
        this.inicio = inicio;
        this.fin = fin;
        this.miParte = miParte;
    }

/**
 * Este metodo es el que ejecuta cada hilo que se crea
 * Realiza la suma de los terminos que van desde el valor "inicial"
 * hasta la cantidad de "terminos"
 * Coloca el resultado en una instancia de la clase Suma
 */
    public void run() {
       double miSuma = 0;
       double alterna;

       alterna = ( 0 == ( inicio % 2 ) )? 4 : -4;
       for ( long i = inicio; i < fin; i++ ) {
           miSuma += alterna/(2 * i + 1);
           alterna *= -1;
       }
       miParte.setSuma( miSuma );
    }

}

public class PiPorSeriesThreads {


/**
 * Realiza las pruebas de la clase
 */
    public static void main( String[] args ) {
        long n = 100000;
        double x = 0;
        Suma miParte[]  = new Suma[ 10 ];	// Una instancia para guardar los resultados de cada hilo
        Thread thread[] = new Thread[ 10 ];	// Una instancia por cada hilo creado
						// Se utilizan para esperar por las sumas de cada hilo

	if ( args.length > 0 ) {
           n = Long.parseLong( args[ 0 ] );
	}
        for ( int i = 0; i < 10; i++ ) {	// Generar los 10 hilos calculadores
           miParte[ i ] = new Suma();
           long inicio = i * n / 10;
           long fin = (i+1)*n/10;
           thread[ i ] = new Thread( new SeriePorPartes( inicio, fin, miParte[ i ] ) );
           thread[ i ].start();
        }

        for ( int i = 0; i < 10; i++ ) {	// Esperar por los resultados de cada hilo
           try {
              thread[ i ].join();
              x += miParte[ i ].getSuma();
//    System.out.println("Valor aportado por hilo " + i + " es " + miParte[ i ].getSuma() );
           } catch ( InterruptedException e) {
           }
        }
        System.out.println( "Valor calculado de PI: " + x );

    }
}

